File: src/ConstantModelOptimizer/TraceSet.cs
//// TRANSLATED FROM EDNA
//using System;
//using System.Collections.Concurrent;
//using System.Collections.Generic;
//using System.IO;
//using System.Linq;
//using System.Threading;
//using System.Threading.Tasks;
//using PacBio.IO;
//using PacBio.Utils;
//using PacBio.Align;
//
//namespace ConstantModelOptimizer
//{
//    public class TraceSet
//    {
//
//        OldCmpH5Reader cmp;
//        BasCollection bas;
//        int maxTraces;
//
//        /// <summary>
//        /// A Basemap 
//        /// </summary>
//        Dictionary<char, int> bm;
//
//        Dictionary<string, float[]> blData;
//
//        Dictionary<string, float[]> sigmaData;
//
//        List<IAlnSummary> Alns;
//
//        public TraceSet(string cmpFile, string fofn, int maximumTraces)
//        {
//            cmp = OldCmpH5Reader.CreateReader (cmpFile);
//            bas = BasCollection.FromFofn (fofn);
//            maxTraces = maxTraces;
//            bm = new Dictionary<char, int> ();
//            for(int i=0; i< bas.BaseMap.Length; i++)
//            {
//                bm [i] = bas.BaseMap [i];
//            }
//            blData = readBaselineData (bm, fofn, out sigmaData);
//            var rng = new RandomCMWC(42);
//            Alns = cmp.Alns.TargetSample ((fun => true), maxTraces, cmp.Alns.Count, rng).ToList ();
//
//        }
//
//        public string ReportsFolder {
//            get { return cmp.ReportsFolder; }
//        }
//
//        public IAlignment GetAlignment(IAlnSummary aln)
//        {
//            return cmp.ReadAlignment (aln);
//        }
//
//        public float[] GetBaselineBias(string id)
//        {
//            blData [id];
//        }
//
//        public float[] GetBaselineSigma(string id)
//        {
//            return sigmaData [id];
//        }
//
//        public int Count {get{return Alns.Length;}}
//
//        public IEnumerable<Trace> Traces() 
//        {
//            foreach (var a in Alns) {
//                var zmwBases = bas.GetRead (a);
//                var tr = new Trace (this, zmwBases, a);
//                yield return tr;
//            }
//
//        }
//
//        public static void readBaseLineData(Dictionary<char, int> bm, string fofnFile, out Dictionary<string, float[]> sgData) 
//        {
//            var resultsDirs = System.IO.File.ReadAllLines (fofnFile).Select (x => Path.GetDirectoryName (x)).Distinct ();
//            var stsFiles = resultsDirs.SelectMany (d => System.IO.Directory.EnumerateFiles (d, "*.sts.csv"));
//
//            var blData = new Dictionary<string, float[]> ();
//            var sgData = new Dictionary<string, float[]>();
//
//            let readStsCsv f =
//                let csvRows = miniCsvReader f
//
//                for r in csvRows do
//                    let zmwId = (r("Movie")).Trim() + "/" + (r("Zmw")).Trim()
//
//                    let bias = Array.zeroCreate 4
//                    bias.[bm.['T']] <- float (r("BaselineLevel_T"))
//                    bias.[bm.['G']] <- float (r("BaselineLevel_G"))
//                    bias.[bm.['A']] <- float (r("BaselineLevel_A"))
//                    bias.[bm.['C']] <- float (r("BaselineLevel_C"))
//                    blData.[zmwId] <- bias
//
//                    let sigm = Array.zeroCreate 4
//                    sigm.[bm.['T']] <- float (r("BaselineStd_T"))
//                    sigm.[bm.['G']] <- float (r("BaselineStd_G"))
//                    sigm.[bm.['A']] <- float (r("BaselineStd_A"))
//                    sigm.[bm.['C']] <- float (r("BaselineStd_C"))
//                    sgData.[zmwId] <- sigm
//                    done
//
//
//                    stsFiles |> Seq.iter readStsCsv
//
//                    (blData, sgData)
//
//
//        }
//    }
//}
//

File: src/Edna/src/Edna/TraceSet.fs
module TraceSet

open System
open PacBio.Utils
open PacBio.IO
open PacBio.Hmm.Utils

open System.Collections.Generic

let miniCsvReader f =
    let headers = (System.IO.File.ReadLines f |> Seq.head).Split ','
    let lines = System.IO.File.ReadLines f |> Seq.skip 1

    let headerIndex = headers |> Seq.mapi (fun idx k -> (k.Trim(), idx)) |> Map.ofSeq

    seq {
        for l in lines do
            let chunks = l.Split ','
            yield (fun k -> chunks.[headerIndex.[k]])
        done }


let readBaselineData (bm : Map<char,int>) fofnFile =
    
    //let convertPath = if Environment.OSVersion.Platform = PlatformID.Win32NT then BasCollection.ConvertFofnPath else id
    let convertPath = id

    let resultsDirs = System.IO.File.ReadAllLines(fofnFile) |> Seq.map (convertPath >> System.IO.Path.GetDirectoryName) |> Seq.distinct
    let stsFiles = resultsDirs |> Seq.collect (fun d -> System.IO.Directory.EnumerateFiles(d, "*.sts.csv"))

    let blData = new Dictionary<string, float[]>()
    let sgData = new Dictionary<string, float[]>()

    let readStsCsv f =
        let csvRows = miniCsvReader f

        for r in csvRows do
            let zmwId = (r("Movie")).Trim() + "/" + (r("Zmw")).Trim()

            let bias = Array.zeroCreate 4
            bias.[bm.['T']] <- float (r("BaselineLevel_T"))
            bias.[bm.['G']] <- float (r("BaselineLevel_G"))
            bias.[bm.['A']] <- float (r("BaselineLevel_A"))
            bias.[bm.['C']] <- float (r("BaselineLevel_C"))
            blData.[zmwId] <- bias

            let sigm = Array.zeroCreate 4
            sigm.[bm.['T']] <- float (r("BaselineStd_T"))
            sigm.[bm.['G']] <- float (r("BaselineStd_G"))
            sigm.[bm.['A']] <- float (r("BaselineStd_A"))
            sigm.[bm.['C']] <- float (r("BaselineStd_C"))
            sgData.[zmwId] <- sigm
        done


    stsFiles |> Seq.iter readStsCsv

    (blData, sgData)


type Trace = { reader : TraceSet; ZmwBases : IZmwBases; alignment : IAlnSummary } with
    member x.ReadId = x.ZmwBases.Zmw.Movie.MovieName + "/" + (string x.ZmwBases.Zmw.HoleNumber)
    member x.HoleNumber = x.ZmwBases.Zmw.HoleNumber
    member x.Metadata = x.ZmwBases.Zmw.Movie
    member x.FullAlignment = x.reader.GetAlignment(x.alignment)
    member x.BaselineBias = x.reader.GetBaselineBias(x.ReadId)
    member x.BaselineSigma = x.reader.GetBaselineSigma(x.ReadId)
    member x.ReportsFolder = x.reader.ReportsFolder

and TraceSet(cmpFile, fofn, ?maxTraces0 : int) =
    let cmp = OldCmpH5Reader.CreateReader(cmpFile)
    let bas = BasCollection.FromFofn(fofn)
    let maxTraces = defaultArg maxTraces0 cmp.Alns.Count

    let bm = bas.BaseMap |> Seq.mapi (fun idx bse -> (bse, idx)) |> Map.ofSeq
    let (blData, sigmaData) = readBaselineData bm fofn

    member x.Alns = 
        // random sample, but preserve order
        let rng = new RandomCMWC(42)
        cmp.Alns.TargetSample((fun _ -> true), maxTraces, cmp.Alns.Count, rng) |> Seq.toList

    member x.Count = x.Alns.Length

    member x.Traces() =
        seq {
          for a in x.Alns do
                let zmwBases = bas.GetRead(a)

                let tr = { reader = x; ZmwBases = zmwBases; alignment = a }
                yield tr
            done }

    member x.ReportsFolder = cmp.ReportsFolder
    member x.GetAlignment(aln) = cmp.ReadAlignment aln
    member x.GetBaselineBias id = blData.[id]
    member x.GetBaselineSigma id = sigmaData.[id]